### ISSUE TITLE
EventProcessorHost not working behind corporate proxy

### ISSUE THREAD

**Issue by PiotrKowalski93** [Azure/azure-event-hubs-dotnet#210]
## Actual Behavior
1. Client is using Microsoft.Azure.EventHubs and EventHubs.Processor. When app is behind corporate proxy there is an exception `Encountered error while fetching the list of EventHub PartitionIds`. 

## Expected Behavior
1. Client is connected to IoT Hub/Event Hub via proxy

## Versions  
- OS platform and version: Windows 10
- .NET Version: .Net Standard
- NuGet package version or commit ID: 
    - Microsoft.Azure.EventHubs 1.0.3
    - Microsoft.Azure.EventHubs.Processor 1.0.3

## Description
I cloned `azure-event-hubs-dotnet` repository and find out that IWebProxy implementation is missing. So I was trying to add handling proxy parameter in connection string.
 I changed `EventHubsConnectionStringBuilder` and `AmqpEventHubClient` like below:

`AmqpEventHubClient`:
```
static TransportSettings CreateWebSocketsTransportSettings(string hostName, string proxyAddress)
{
     var proxy = new Proxy(new Uri(proxyAddress));

     //[UriBuilder]
      var ts = new WebSocketTransportSettings()
      {
        Uri = uriBuilder.Uri,
        Proxy = proxy,
      };
      return ts;
}
```
`EventHubsConnectionStringBuilder`:

```
void ParseConnectionString(string connectionString)
{
     ...
     else if (key.Equals(WebProxy, StringComparison.OrdinalIgnoreCase))
     {
        this.WebProxyAddress = value;
     }
     ...
}
```

```
public override string ToString()
{
    ...
   if (!string.IsNullOrWhiteSpace(this.WebProxyAddress))
            {
                connectionStringBuilder.Append($"{WebProxy}{KeyValueSeparator}{this.WebProxyAddress}{KeyValuePairDelimiter}");
            }
    ...
}
```
Properties and variables:
```
static readonly string WebProxy = "ProxyAddress";

/// <summary>
/// Proxy addres for Event Hub Client
/// </summary>
public string WebProxyAddress { get; set; }
```

My connection string containg proxy adress:
private const string EhConnectionString = `"Endpoint=sb://xxxxx.xxx.xxx/;SharedAccessKeyName=xxxxxx;SharedAccessKey=xxxx;ProxyAddress=http://xxx.xxx.xxx.xxx:xx";`

This still does not work. Im in touch with our IT Admins because One issue can be that port on our proxy is blocked for websocket. But other it might be that proxy is not handled properly in the code.

1. Are you working on that ? Do you have any tips to handle that?
2. How is connection working with azure? In wireshark I saw two separate requests for HTTP and TCP. For which I should set up ssh tunneling to achieve proper connection with my IoT Hub/Event Hub? Can I use localhost in my connection string when I use port forwarding via PuTTy?

**Comment by serkantkaraca** [comment]
There was a recent proxy fix on AMQP liblary - https://github.com/Azure/azure-amqp/commit/36543207a086b7468d702b66b60f95b1a729cf4f

Can you try building the client with most recent AMQP release?

**Comment by serkantkaraca** [comment]
We are starting to work on proxy configuration support and this will be available in the next client release.

**Comment by dimitertodorov** [comment]
Hey, I just compiled from the latest code. 
Proxy settings do not appear to take effect when using netcoreapp2.0 as target; either by picking up system proxy or manually in code.
e.g.
```
var webProxy = new WebProxy("http://5.6.7.3:3128", false);
webProxy.Credentials = new NetworkCredential("asdsadsad","asdsadsad");
eventHubClient.WebProxy = webProxy;
```

Using the full 4.6.1 framework appears to pick up system proxy settings properly, and allows for setting from code as well.

Is there a chance you could release a new version (or preview) into NuGet.

**Comment by serkantkaraca** [comment]
Net Standard 2.0 SDK currently doesn't support proxy. AFAIK, they are fixing it soon. Once the fix is out we will move to this new SDK and it should start working after then.

**Comment by serkantkaraca** [comment]
Closing since this will be addressed by core Fx.

**Comment by alexmg** [comment]
@serkantkaraca Do you have a link to the CoreFX issue for the proxy? I am guessing it is https://github.com/dotnet/corefx/issues/30166.

I am running the `EventHubProcessor` on `netcoreapp2.1` and it is during the initialization of `AzureStorageCheckpointLeaseManager` that I see proxy related failures.

The `PartitionManager` is using the `IWebProxy` provided to it via the `EventProcessorOptions` when creating the `EventHubClient` instance, but the call to `AzureStorageCheckpointLeaseManager.Initialize` happens prior to that in `EventProcessorHost.RegisterEventProcessorFactoryAsync`. The `EventProcessorOptions` have not even been set on the `EventProcessorHost` that is passed to the `Initialize` method.

It seems that proxy support is incomplete without the blob client also supporting the proxy correctly and https://github.com/Azure/azure-storage-net/issues/711 suggests this is not the case. I don't actually see an API that would allow you to pass the `IWebProxy` instance through to `CloudBlobClient`. I can connect to the event hub directly with `EventHubClient` after settings the `IWebProxy` as you are doing in your code. It is the `CloudBlobClient` not using the same instance that seems to be the problem.

**Comment by serkantkaraca** [comment]
Thanks for reporting this. I will check on couple things and see how to wire up proxy settings to storage client.

**Comment by serkantkaraca** [comment]
I have created a PR for .Net 4.6 build. Once storage client is ready, we will wire this up for Standard client as well.

https://github.com/Azure/azure-event-hubs-dotnet/pull/311

**Comment by mathavanmani** [comment]
Hi @serkantkaraca , Looking at the latest code https://github.com/Azure/azure-sdk-for-net/blob/166356f5367ef208f27a096652148b886019a053/sdk/eventhub/Microsoft.Azure.EventHubs.Processor/src/AzureStorageCheckpointLeaseManager.cs, it looks like it is fixed only for .net full framework. Any plans to apply this fix for .Net standard/Core as well?

**Comment by serkantkaraca** [comment]
This repository is merged in to central SDK repository. Can you create a new issue on https://github.com/Azure/azure-sdk-for-net ? Let's track this there.

**Comment by mathavanmani** [comment]
Hi @serkantkaraca , Thanks for your comment. Raised a new one https://github.com/Azure/azure-sdk-for-net/issues/8849
